src/Operation/Init.php

```diff
@@ -29,22 +31,29 @@ final class Init extends AbstractOperation
      */
     public function __invoke(): Closure
     {
-        return
+        $callback =
+            /**
+             * @param T $value
+             * @param TKey $key
+             * @param CachingIterator<TKey, T> $iterator
+             */
+            static fn ($value, $key, CachingIterator $iterator): bool => $iterator->getInnerIterator()->valid();
```

Comment:
for some reason Psalm thinks that `getInnerIterator` returns `Traversable` here instead of `Iterator`, but I cannot replicate it in the playground: https://psalm.dev/r/037fdbb040. The trace shows `Iterator` there.

But in this project:
```
INFO: Trace - src/Operation/Init.php:42:17 - $inner: Traversable (see https://psalm.dev/224)
                /** @psalm-trace $inner */
                $inner = $iterator->getInnerIterator();
```

Could it be the Psalm version? (we must be using an older one)

Reply:
Haha ! I almost made the same Psalm example to try it out on my own :)

I don't know where the issue comes from yet, I'm investigating!

Reply:
It only happens with PHP 8. Weird.

Reply:
I think I should submit this to PSalm.

![image](https://user-images.githubusercontent.com/252042/130579482-84eb095c-4751-4f59-99b4-131ff445a29c.png)

Reply:
Weird. Using `CachingIterator::hasNext()` fixed the issue.

Reply:
I think we should open an issue on Psalm with that, sounds like a problem in PHP 8

Reply:
I did here: https://github.com/vimeo/psalm/pull/6359 we'll see how it goes.